### PR TITLE
Allow configuring `AuthorizeSiteAccess` plug site param

### PR DESCRIPTION
**Pull request by apata**
### Changes

- Allow auth site access plug to specify that domain is found in request body at some key, needed for api/docs/query endpoint where :domain or :website does not appear in path params
- Nuisance: Stop typescript --watch command clearing previous server stdout when running make server

### Tests
- [x] Automated tests have been added

### Changelog
- [x] This PR does not make a user-facing change

### Documentation
- [x] This change does not need a documentation update

### Dark mode
- [x] This PR does not change the UI

### CODE REVIEW

File: lib/plausible_web/plugs/authorize_site_access.ex [hunk 1]
```diff
@@ -63,12 +76,16 @@ defmodule PlausibleWeb.Plugs.AuthorizeSiteAccess do
...
+  defp get_domain_from_conn(conn, nil) do
+    conn.path_params["domain"] || conn.path_params["website"]
+  end
+
+  defp get_domain_from_conn(conn, conn_params_domain_accessor) do
+    conn.params[conn_params_domain_accessor]
+  end
+
+  defp get_site_with_role(conn, current_user, conn_params_domain_accessor) do
+    domain = get_domain_from_conn(conn, conn_params_domain_accessor)
```
Comment (zoldar) [inline lib/plausible_web/plugs/authorize_site_access.ex:88]: When the param is not provided at all (`site_id`, for instance), the plug will crash when executing the query, AFAICT. Is `site_id` optional here? If not, perhaps we should return 404, like in other cases?

File: lib/plausible_web/plugs/authorize_site_access.ex [hunk 2]
```diff
@@ -22,10 +22,23 @@ defmodule PlausibleWeb.Plugs.AuthorizeSiteAccess do
     allowed_roles
   end
 
-  def call(conn, allowed_roles) do
+  def init({allowed_roles, conn_params_domain_accessor})
```
Comment (zoldar) [inline lib/plausible_web/plugs/authorize_site_access.ex:25]: IMHO explicit map instead of a tuple is better for readability.

File: lib/plausible_web/plugs/authorize_site_access.ex [hunk 3]
```diff
@@ -12,7 +12,7 @@ defmodule PlausibleWeb.Plugs.AuthorizeSiteAccess do
 
   def init([]), do: @all_roles
 
-  def init(allowed_roles) do
+  def init(allowed_roles) when is_list(allowed_roles) do
```
Comment (zoldar) [inline lib/plausible_web/plugs/authorize_site_access.ex:15]: 👍

File: lib/plausible_web/plugs/authorize_site_access.ex [hunk 4]
```diff
@@ -22,10 +22,23 @@ defmodule PlausibleWeb.Plugs.AuthorizeSiteAccess do
     allowed_roles
   end
 
-  def call(conn, allowed_roles) do
+  def init({allowed_roles, conn_params_domain_accessor})
+      when is_list(allowed_roles) and is_binary(conn_params_domain_accessor) do
```
Comment (zoldar) [inline lib/plausible_web/plugs/authorize_site_access.ex:26]: Maybe `domain_param`? Or even `site_param`?

File: lib/plausible_web/plugs/authorize_site_access.ex [hunk 5]
```diff
@@ -22,10 +22,23 @@ defmodule PlausibleWeb.Plugs.AuthorizeSiteAccess do
...
+  def call(conn, options) do
     current_user = conn.assigns[:current_user]
 
-    with {:ok, %{site: site, role: membership_role}} <- get_site_with_role(conn, current_user),
+    {allowed_roles, conn_params_domain_accessor} =
+      if is_tuple(options) do
+        options
+      else
+        {options, nil}
+      end
```
Comment (zoldar) [inline lib/plausible_web/plugs/authorize_site_access.ex:38]: Instead of pattern matching in `call` we could return consistent/normalized args from `init` already (via map?)

Comment (apata) [reply lib/plausible_web/plugs/authorize_site_access.ex:38]: Ah, I tried that but then that would have me refactor the args to AuthoriseSiteAccess.call

File: lib/plausible_web/plugs/authorize_site_access.ex [hunk 6]
```diff
@@ -63,12 +76,16 @@ defmodule PlausibleWeb.Plugs.AuthorizeSiteAccess do
     end
   end
 
-  defp get_site_with_role(conn, current_user) do
-    # addition is flimsy, do we need an extra argument on plug init to control where we look for the domain?
-    domain =
-      conn.path_params["domain"] || conn.path_params["website"] ||
-        (conn.method == "POST" && conn.path_info == ["api", "docs", "query"] &&
-           conn.params["site_id"])
```
Comment (ruslandoga) [inline lib/plausible_web/plugs/authorize_site_access.ex:71]: 👋 

Was just passing through and saw this comment :) I wonder if we can use `:assings` on scope or route for the relevant paths and set some value there and then based on that value perform different lookups?

Alternatively, if the logic in this plug can be split in two:
- extract site plug
- authorize site plug

Comment (zoldar) [reply lib/plausible_web/plugs/authorize_site_access.ex:71]: Passing anonymous function to plug config is something I've been fiddling with initially, but then I got reminded `init` is evaluated at compile time, and supporting passing stuff like anonymous functions at compile time is AFAIK quite a mess 😬 (like in case on [abinsthe](https://elixirforum.com/t/escaping-anonymous-function-in-module-attribute/60664/2)).

The scope with `private` sounds neat but it also makes the configration a bit too replaced from the actual use site - and the relationship not obvious. IMHO keeping config close to the plug is better from maintenance perspective - even if the failure mode is relatively safe (missing/stale parameter would just cause rejection, no accidentally allowing the request).

Comment (ruslandoga) [reply lib/plausible_web/plugs/authorize_site_access.ex:71]: > The scope with private sounds neat but it also makes the configration a bit too replaced from the actual use site - and the relationship not obvious.

In my mid it's the opposite -- the use site is the router.

File: test/plausible_web/controllers/api/internal_controller/docs_query_test.exs [hunk 7]
```diff
@@ -1,5 +1,5 @@
 defmodule PlausibleWeb.Api.InternalController.DocsQueryTest do
-  use PlausibleWeb.ConnCase, async: true
+  use PlausibleWeb.ConnCase, async: false
```
Comment (aerosol) [inline test/plausible_web/controllers/api/internal_controller/docs_query_test.exs:2]: what's the reason for this change?

Comment (aerosol) [reply test/plausible_web/controllers/api/internal_controller/docs_query_test.exs:2]: ah, patch env. welp.

Comment (aerosol) [reply test/plausible_web/controllers/api/internal_controller/docs_query_test.exs:2]: wait, there's no env patching in that test suite

**Review by aerosol — APPROVED** [review]

File: test/plausible_web/controllers/api/internal_controller/docs_query_test.exs (hunk 7 above)
Comment (zoldar) [reply test/plausible_web/controllers/api/internal_controller/docs_query_test.exs]: https://github.com/plausible/analytics/pull/4597/commits/61be18219040275f9c5e70a27d1155f5385b6a69
